src/components/legend.ts

```diff
@@ -237,88 +514,107 @@ namespace Plottable.Components {
         return [];
       }
 
-      let entities: Entity<Legend>[] = [];
-      let layout = this._calculateLayoutInfo(this.width(), this.height());
-      let legendPadding = this._padding;
-      let legend = this;
-      this.content().selectAll("g." + Legend.LEGEND_ROW_CLASS).each(function(d: any, i: number) {
-        let lowY = i * layout.textHeight + legendPadding;
-        let highY = (i + 1) * layout.textHeight + legendPadding;
-        let symbolY = (lowY + highY) / 2;
-        let lowX = legendPadding;
-        let highX = legendPadding;
-        d3.select(this).selectAll("g." + Legend.LEGEND_ENTRY_CLASS).each(function(value: string) {
-          highX += layout.entryLengths.get(value);
-          let symbolX = lowX + layout.textHeight / 2;
-          if (highX >= p.x && lowX <= p.x &&
-              highY >= p.y && lowY <= p.y) {
-            let entrySelection = d3.select(this);
-            let datum = entrySelection.datum();
-            entities.push({
-              datum: datum,
-              position: { x: symbolX, y: symbolY },
-              selection: entrySelection,
-              component: legend,
-            });
+      function within(p: Point, bounds: Bounds) {
+        return bounds.topLeft.x <= p.x
+          && bounds.bottomRight.x >= p.x
+          && bounds.topLeft.y <= p.y
+          && bounds.bottomRight.y >= p.y;
+      };
+
+      const table = this._buildLegendTable(this.width(), this.height());
+      return table.reduce((entity: Entity<Legend>[], row: LegendRow, rowIndex: number) => {
+        if (entity.length !== 0) {
+          // we've already found the nearest entity; just return it.
+          return entity;
+        }
+
+        const rowBounds = table.getBounds(row);
+        const withinRow = within(p, rowBounds);
+
+        if (!withinRow) {
+          // the nearest entity isn't within this row, continue;
+          return entity;
```

Comment:
since we've gotten here without modifying `entity` and given the first conditional, this is always going to be an empty array right? Using `reduce` in this way seems a little odd to me. Why even use an array as opposed to value/null?